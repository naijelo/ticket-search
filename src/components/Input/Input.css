.input {
    width: 230px;
}

.medium {height: 23px;}
.large {height: 50px;}