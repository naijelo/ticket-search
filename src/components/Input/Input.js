import React from "react";
// import "./Input.css";

const Input = ({variant = "medium", children, label, ...rest}) => {

    return (
        <div>
            <label>{label}
                <input type="text" 
                    className={`input ${variant}`} 
                    required 
                    {...rest}/>
            </label>
        </div>
    );
}

export default Input;