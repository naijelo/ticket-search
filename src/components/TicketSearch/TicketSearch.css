.ticketSearch {
    margin: 0 auto;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.inputWrapper {
    display: flex;
    justify-content: space-between;
}

.inputWrapper__from {
    display: flex;
    flex-direction: column;
}
.datepicker {
    margin-top: 50px;
}