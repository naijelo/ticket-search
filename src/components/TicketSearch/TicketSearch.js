import React, {useState, useEffect} from "react";

import { StyledButton } from "../Button/Button.style.js";
import { StyledInput } from "../Input/Input.style.js";
import { StyledPagination } from "../Pagination/Pagination.style.js";
import { StyledDropDown } from "../DropDown/DropDown.style.js";
import { StyledCards } from "../Cards/Cards.style.js";
import Calendar from "../Calendar/Calendar";

import * as axios from "axios";
import jsondata from '../Input/cities.json';

import styled from "styled-components";

const TicketWrapper = styled.form `
    margin: 0 auto;
    display: flex;
    flex-direction: column;
    align-items: center;
    @media (max-width: 575px) {
        justify-content: start;
    }
`
const InputWrapper = styled.div `
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    margin: 0 auto;
    @media (max-width: 575px) {
        flex-direction: column;
    }
`

const InputSubWrapper = styled.div`
    display: flex;
    flex-direction:column;
`

const TicketSearch = (props) => {

    
    // const proxy = "https://cors-anywhere.herokuapp.com/";
    // const token = "e44551a4aa602315a353a5fc2d7bed15";
    // const api = "http://api.travelpayouts.com/data/ru/cities.json";

    //errors
    const [dataError, setDataError] = useState(false);
    //set data into inputs
    const [inputFrom, setInputFrom] = useState("");
    const [inputTo, setInputTo] = useState("");
    
    //get date from inputs
    const [inputDate, setInputDate] = useState("");
    const [fullDate, setFullDate] = useState("");
    
    //filter state and arrow of data from server
    const [filteredDataFrom, setFilteredDataFrom] = useState([]);
    const [filteredDataTo, setFilteredDataTo] = useState([]);
    const [cities, setCities] = useState([]);
    
    const [ticketData, setTicketData] = useState([]);
    const [loading, setLoading] = useState(false);
    const [currentPage, setCurrentPage] = useState(1);
    const [postsPerPage] = useState(10);

    //activate drop down menu from button
    const [dateVisibility, setDateVisibility] = useState(false);
    
    //paginator support data
    const indexOfLastCard = currentPage * postsPerPage;
    const indexOfFirstCard = indexOfLastCard - postsPerPage;
    const currentCards = ticketData.slice(indexOfFirstCard, indexOfLastCard);

    const paginate = (pageNumber) => {
        setCurrentPage(pageNumber);
    }
    
    //fetch cities from aviasales api using useEffect cause of sideeffect
    useEffect(() => {
        const fetchData = async () => {
            setLoading(true);
            const response = await axios.get("/cities.json");
            setCities(response.data);
            if (!response) {
                setCities(jsondata);
            }
            setLoading(false);
        }
        fetchData();
    }, []);
    
    //implement filter to find matches in response and input
    const handleFilter = (event, setFilteredData, setWordEntered) => {
        
        if (event.target.value === "") {
            setFilteredData([]);
        } else {
            setFilteredData(cities.filter((value) => {
                return value.name.toLowerCase().includes(event.target.value.toLowerCase());
            }));
        }
        setWordEntered(event.target.value);
    };

    //make object from received data and send it to server
    const handleSubmit = (event) => {
        event.preventDefault();
        const formData = {
            "from": cities.find((item) => inputFrom === item.name).code,
            "to": cities.find((item) => inputTo === item.name).code,
            "date": fullDate,
        }
        console.log(formData);
        const requestData =`?origin=${formData.from}&destination=${formData.to}&depart_day=${formData.date}&one_way=true`;

        axios.get("http://min-prices.aviasales.ru/calendar_preload" + requestData)
        .then((response) => {
            response = response.data.best_prices;
        console.log(response);
        setTicketData(response)
        })
        .catch(err => {
            setDataError(true);
        });
    };

    //add city from dropdown list to input
    const addCity = (event, addCityToState, setFilteredData) => {
        const target = event.target;
        if (target.tagName.toLowerCase() === "li") {
            setFilteredData([]);
            addCityToState(target.textContent);
        }
    }

    //get date from date-input
    const handleInput = (date) => {
        const fullDateTemp = `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`;
        setFullDate(fullDateTemp);
        setInputDate(date);
    }

    //visibility of calendar
    let changeVisibility = () => setDateVisibility(!dateVisibility);

    return (
        <div>
            <TicketWrapper
                onSubmit={handleSubmit}
                >
                <InputWrapper>
                    <InputSubWrapper>
                    <StyledInput label={"Откуда"} 
                        value={inputFrom}
                        onInput={event => handleFilter(event, setFilteredDataFrom, setInputFrom)} 
                    />
                    
                    {filteredDataFrom.length !== 0 && 
                        <StyledDropDown setCustomInput={setInputFrom} 
                            filteredData={filteredDataFrom} 
                            setFilteredData={setFilteredDataFrom}
                            addCity={addCity}/>}
                    </InputSubWrapper>

                    <InputSubWrapper>
                    <StyledInput label={"Куда"} 
                        value={inputTo}
                        onInput={event => handleFilter(event, setFilteredDataTo, setInputTo)}
                    />
                    {filteredDataTo.length !== 0 && 
                        <StyledDropDown setCustomInput={setInputTo} 
                            filteredData={filteredDataTo} 
                            setFilteredData={setFilteredDataTo}
                            addCity={addCity}/>}
                    </InputSubWrapper>

                </InputWrapper>
                    <div onClick={changeVisibility}>{fullDate ? `Date:${fullDate}` : "Выберите дату"}</div>
                    {dateVisibility === true && 
                    <Calendar onClick={handleInput}
                              setFullDate={setFullDate}
                              setDateVisibility={setDateVisibility}
                    />}

                    <StyledButton title="Submit" 
                        children={"Поиск"}
                    />
            </TicketWrapper>

            <StyledCards ticketData={currentCards} 
                    loading={loading}
                    dataError={dataError}/>
            <StyledPagination postsPerPage={postsPerPage} totalPosts={ticketData.length} paginate={paginate}/>

        </div>
    )
}

export default TicketSearch;