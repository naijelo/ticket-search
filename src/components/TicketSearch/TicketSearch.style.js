import styled from "styled-components";


export const MainTicketSearchWrapper = styled.form `
    margin-top: 10vh;
    padding: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    @media (max-width: 575px) {
        justify-content: start;
        align-items: center;
        flex-direction: column;
    }
`

export const InputSubWrapper = styled.div`
    display: flex;
    flex-direction: column;
    margin-right: 2px;
    width: 20vw;
    height: 40px;
    margin-right: 10px;
    @media (max-width: 575px) {
       margin-bottom: 10px;
       width: 50%;
       margin-right: 0;
    }
`

export const DatePickerActivateField = styled.div `
    color: orange;
    font-weight: 600;
    position: relative;
    background-color: #fff;
    cursor: pointer;
    width: 20vw;
    height: 40px;
    border-radius: 4px;
    margin-right: 1px;
    text-align: center;
    line-height: 35px;
    @media (max-width: 575px) {
       margin-right: 0;
    }
`
export const CalendarWrapper = styled.div`
    display: flex;
    flex-direction: column;
`
