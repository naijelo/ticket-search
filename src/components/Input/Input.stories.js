import React from "react";
import Input from "./Input";

export default {
    title: "Input",
    component: Input
}

export const Medium = () => <Input variant="medium" placeholder="medium"></Input>
export const Large = () => <Input variant="large" placeholder="large"></Input>