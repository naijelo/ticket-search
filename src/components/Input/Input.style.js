import styled from "styled-components";
import Input from "./Input";

export const StyledInput = styled(Input)`
    position: relative;
    border-radius: 4px;
    width: 100%;
    height: 38px;
    outline: none;
    border: none;
`