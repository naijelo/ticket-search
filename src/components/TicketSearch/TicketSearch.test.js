import React from 'react';
import { shallow, render, mount } from "enzyme";
import sinon from 'sinon';
import initStoryshots from '@storybook/addon-storyshots';

import TicketSearch from "./TicketSearch";
import { DatePickerActivateField } from "./TicketSearch.style.js";
import { DatePicker } from "../Calendar/Calendar.style.js";

import { StyledInput } from '../Input/Input.style';

initStoryshots({});

let wrapper;
test('renders a div into TicketSearch', () => {
  let wrapper = shallow(<TicketSearch />);
  expect(wrapper.find('div'));
});

test("render of TicketSearch", () => {
  expect(wrapper).not.toBenull;
});

describe("inputs", () => {
  const props = {
    onInput: jest.fn()
  };

  beforeEach(() => {
    wrapper = mount(<TicketSearch {...props}/>)
  });

  test("correctly number of inputs", () => {
    expect(wrapper.find("input").length).toEqual(2);
  });

  test("correctly from-input behavior", () => {
    wrapper.find("input").at(0).simulate("change", { target: { value: "foo" }});
  });

  test("correctly to-input behavior", () => {
    wrapper.find("input").at(1).simulate("change", { target: { value: "foo" }});
  });

  test('should have proper props for input field', () => {
    expect(wrapper.find('input').at(0).props()).toEqual({
      onInput: expect.any(Function),
      value: "",
      placeholder: expect.any(String),
      className: expect.any(String),
      autoFocus: true,
      type: "text",
      required: expect.any(Boolean),
    });
  });

  test("test input onChange", () => {
    const mockCallBack = jest.fn();
    const input = mount(<StyledInput onInput={mockCallBack}/>);
    input.find("input").simulate("input");
    expect(mockCallBack).toHaveBeenCalled();
  });

  test("first input have autofocus by default", () => {
    // const handleFilter = jest.fn();
    const styledInput = wrapper.find(StyledInput);
    expect(styledInput.at(0).prop('autoFocus')).toBeTruthy();
  });

  test("handreFilter work on first input while change value", () => {
    const handleFilter = jest.fn();
    const styledInput = wrapper.find(StyledInput);
    styledInput.at(0).find("input").simulate('input');
    expect(handleFilter.mock.calls.length).toEqual(0);

  });

  test("handreFilter work on second input while change value", () => {
    const handleFilter = jest.fn();
    const styledInput = wrapper.find(StyledInput);
    styledInput.at(1).find("input").simulate('input');
    expect(handleFilter.mock.calls.length).toEqual(0);
  });

 


});

describe("calendar", () => {
  let wrapper;
  const props = {
    onClick: jest.fn()
  };
  beforeEach(() => {
      wrapper = mount(<TicketSearch {...props}/>)
    });

  test("render of calendar activate button", () => {
    expect(wrapper.find("DatePickerActivateField")).not.toBenull;
  });

  test("dont see dropdown before click", () => {
    expect(wrapper.find("DatePicker").length).toEqual(0);
  });

  test("DatePickerActivateField is defined", () => {
    expect(wrapper.find(<DatePickerActivateField/>)).toBeDefined();
  });

  test("div in DatePickerActivateField", () => {
    const DatePickerActivate = wrapper.find(DatePickerActivateField);
    expect(DatePickerActivate.find("div")).toBeDefined();
  });

  test("don't see dropdown before click", () => {
    const DatePickerActivate = wrapper.find(DatePickerActivateField);
    expect(DatePickerActivate.find("div")).toBeDefined();
  });
  
  test("see dropdown after click", () => {
    const DatePickerActivate = wrapper.find(DatePickerActivateField);
    DatePickerActivate.find("div").simulate("click");
    expect(wrapper.find(DatePicker)).toBeDefined();
  });

});

describe("button", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<TicketSearch/>)
  });

  test("render a button on the page", () => {
    expect(wrapper.find("button").toBeDefined);
  })
})


