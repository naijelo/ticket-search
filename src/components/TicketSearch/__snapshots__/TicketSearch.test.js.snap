// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Button Primary 1`] = `
<button
  className="sc-eCApnc hlZrKE"
  inputColor="#ffa928"
>
  Primary
</button>
`;

exports[`Storyshots Button Secondary 1`] = `
<button
  className="sc-eCApnc kuZrco"
  inputColor="grey"
>
  Secondary
</button>
`;

exports[`Storyshots DropDown Dark Theme 1`] = `
<ul
  className="sc-iCoGMd brSBGG"
  dropdownBackColor="black"
  dropdownColor="white"
>
  <li
    className="from"
    onClick={[Function]}
  >
    Москва
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Минск
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Владивосток
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Кострома
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Канберра
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Канберра
  </li>
</ul>
`;

exports[`Storyshots DropDown White Theme 1`] = `
<ul
  className="sc-iCoGMd iUbvNu"
  dropdownBackColor="white"
  dropdownColor="black"
>
  <li
    className="from"
    onClick={[Function]}
  >
    Москва
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Минск
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Владивосток
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Кострома
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Канберра
  </li>
  <li
    className="from"
    onClick={[Function]}
  >
    Канберра
  </li>
</ul>
`;

exports[`Storyshots Input Large 1`] = `
<div>
  <label>
    <input
      className="sc-jSFjdj hlrREt"
      inputHeight="76px"
      required={true}
      type="text"
    />
  </label>
</div>
`;

exports[`Storyshots Input Standart 1`] = `
<div>
  <label>
    <input
      className="sc-jSFjdj hjfOvn"
      inputHeight="38px"
      required={true}
      type="text"
    />
  </label>
</div>
`;

exports[`Storyshots Pagination Dark Pagination 1`] = `
<nav
  className="sc-gKAaRy facYno"
  pagiBGColor="black"
  pagiColor="white"
>
  <ul
    className="pagination"
  >
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        5
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        6
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        7
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        8
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        9
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Storyshots Pagination Light Pagination 1`] = `
<nav
  className="sc-gKAaRy eSLWXE"
  pagiBGColor="white"
  pagiColor="black"
>
  <ul
    className="pagination"
  >
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        5
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        6
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        7
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        8
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        9
      </a>
    </li>
    <li
      className="pageItem"
    >
      <a
        className="pageLink"
        href="!#"
        onClick={[Function]}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;
